### PR TITLE
ARROW-8401: [C++] Add byte-stream-split AVX2/AVX512 implementation

### PR DESCRIPTION
**Pull request by frankdjx**
For decode, similar implementation with the SSE version for both AVX512 and AVX2.

For encode, float path implemented for AVX2/AVX512 also double for AVX512.
AVX2 double fall back to SSE as currently no epi16 permute support.

BM_ByteStreamSplit result reach up to 3x for AVX512, 2x for AVX2.

Signed-off-by: Frank Du <frank.du@intel.com>

### CODE REVIEW

**Comment by jianxind** [comment]
Below is the size 1024 result,

SSE:
BM_ByteStreamSplitDecode_Float_SIMD/1024            132 ns          132 ns      5296552 bytes_per_second=28.891G/s
BM_ByteStreamSplitDecode_Double_SIMD/1024           387 ns          386 ns      1818101 bytes_per_second=19.7603G/s
BM_ByteStreamSplitEncode_Float_SIMD/1024            263 ns          263 ns      2663415 bytes_per_second=14.5051G/s

AVX2:
BM_ByteStreamSplitDecode_Float_SIMD/1024            104 ns          104 ns      6727450 bytes_per_second=36.686G/s
BM_ByteStreamSplitDecode_Double_SIMD/1024           265 ns          265 ns      2641345 bytes_per_second=28.7856G/s
BM_ByteStreamSplitEncode_Float_SIMD/1024            169 ns          169 ns      4147448 bytes_per_second=22.6001G/s

AVX512:
BM_ByteStreamSplitDecode_Float_SIMD/1024           70.3 ns         70.2 ns      9315691 bytes_per_second=54.356G/s
BM_ByteStreamSplitDecode_Double_SIMD/1024           173 ns          173 ns      4038812 bytes_per_second=44.0386G/s
BM_ByteStreamSplitEncode_Float_SIMD/1024           88.3 ns         88.2 ns      7936455 bytes_per_second=43.2519G/s

File: cpp/src/arrow/util/byte_stream_split.h [hunk 1]
```diff
@@ -171,6 +171,381 @@ void ByteStreamSplitEncodeSSE2(const uint8_t* raw_values, const size_t num_value
...
+                              &output_data[(i * kNumStreams + j) * sizeof(__m256i)]),
+                          final_result[j]);
+    }
+  }
+}
+
+template <typename T>
+void _ByteStreamSplitEncodeAVX2(const uint8_t* raw_values, const size_t num_values,
+                                uint8_t* output_buffer_raw) {
+  constexpr size_t num_streams = sizeof(T);
```
Comment (emkornfield) [inline cpp/src/arrow/util/byte_stream_split.h:252]: style: kNumStreams

File: cpp/src/arrow/util/byte_stream_split.h [hunk 2]
```diff
@@ -171,6 +171,381 @@ void ByteStreamSplitEncodeSSE2(const uint8_t* raw_values, const size_t num_value
...
+    for (size_t i = 0; i < num_streams; ++i) {
+      _mm512_storeu_si512(&output_buffer_streams[i][block_index], final_result[i]);
+    }
+  }
+}
+#endif
+
+#if defined(ARROW_HAVE_SIMD_SPLIT)
+template <typename T>
+void inline ByteStreamSplitDecodeSIMD(const uint8_t* data, int64_t num_values,
```
Comment (emkornfield) [inline cpp/src/arrow/util/byte_stream_split.h:521]: Style
```suggestion
void inline ByteStreamSplitDecodeSimd(const uint8_t* data, int64_t num_values,
```

File: cpp/src/arrow/util/byte_stream_split.h [hunk 3]
```diff
@@ -171,6 +171,381 @@ void ByteStreamSplitEncodeSSE2(const uint8_t* raw_values, const size_t num_value
...
+  return _ByteStreamSplitDecodeAVX2(data, num_values, stride, out);
+#elif defined(ARROW_HAVE_SSE4_2)
+  return _ByteStreamSplitDecodeSSE2(data, num_values, stride, out);
+#else
+#error "ByteStreamSplitDecodeSIMD not defined"
+#endif
+}
+
+template <typename T>
+void inline ByteStreamSplitEncodeSIMD(const uint8_t* raw_values, const size_t num_values,
```
Comment (emkornfield) [inline cpp/src/arrow/util/byte_stream_split.h:535]: SIMD->Simd

File: cpp/src/arrow/util/byte_stream_split.h [hunk 4]
```diff
@@ -171,6 +171,381 @@ void ByteStreamSplitEncodeSSE2(const uint8_t* raw_values, const size_t num_value
 
 #endif
 
+#if defined(ARROW_HAVE_AVX2)
+template <typename T>
+void _ByteStreamSplitDecodeAVX2(const uint8_t* data, int64_t num_values, int64_t stride,
```
Comment (emkornfield) [inline cpp/src/arrow/util/byte_stream_split.h:176]: style: I think it would be better to move all SIMD implementation to a new header file  called something like  byte_stream_split_simd_internal.h and move put the functions in an "internal" namespace instead of prefixing with and underscore.

**Review by emkornfield — CHANGES_REQUESTED** [review]
Can you add unit tests that test the different AVX implementation directly, its not clear that there is good unit test coverage at the moment.

**Comment by wesm** [comment]
You also probably know this, but the BYTE_STREAM_SPLIT encoding is brand new to the project and so not too many people will be able to reap the benefits of these optimizations. Wanted to mention anyway

File: cpp/src/arrow/util/byte_stream_split.h (hunk 1 above)
Comment (jianxind) [reply cpp/src/arrow/util/byte_stream_split.h:252]: done

**Comment by jianxind** [comment]
> Can you add unit tests that test the different AVX implementation directly, its not clear that there is good unit test coverage at the moment.

More tests added for different block and with suffix, and also fix for other comments. Pls help to review it again. Thanks.

**Comment by jianxind** [comment]
> You also probably know this, but the BYTE_STREAM_SPLIT encoding is brand new to the project and so not too many people will be able to reap the benefits of these optimizations. Wanted to mention anyway

Thanks, actually I don't know as I am very new to arrow:) And it will be very great if you can input us some priority lists of encoding which are more popular to end user, or where I can find the info. We can check the candidates though not each encoding may suitable to SIMD.

**Comment by wesm** [comment]
Cool, we're happy to have a SIMD expert involved in the project! Maybe we can have a mailing list discussion about the SIMD optimization topic since many people may have ideas about areas where there may be high value. One idea is SIMD-accelerated small size hashes, for values under say 128 bytes I wonder if we can do better than xxhash3

File: cpp/src/parquet/encoding_test.cc [hunk 5]
```diff
@@ -1096,10 +1096,16 @@ TYPED_TEST(TestByteStreamSplitEncoding, BasicRoundTrip) {
   // require a scalar loop for the suffix.
 
   // Exercise only the SIMD loop.
-  ASSERT_NO_FATAL_FAILURE(this->Execute(256, 1));
+  ASSERT_NO_FATAL_FAILURE(this->Execute(16, 1));   // 1 __m128i for SSE
```
Comment (emkornfield) [inline cpp/src/parquet/encoding_test.cc:1099]: it can help readability to label constants with /*parameter_name = */

File: cpp/src/arrow/util/sse_util.h [hunk 6]
```diff
@@ -32,8 +32,15 @@
 #include <nmmintrin.h>
 #endif
 
+// Enable the SIMD for ByteStreamSplit Encoder/Decoder
+#define ARROW_HAVE_SIMD_SPLIT
```
Comment (emkornfield) [inline cpp/src/arrow/util/sse_util.h:36]: this seems superfluous, I think we can keep using the ARROW_HAVE_SSE4_2 where this is used, since that is the lowest level instruction set that is supported?  If you feel it is useful, I think this should be moved to the the new header file you introduced instead of the generic utilities.

File: cpp/src/arrow/util/sse_util.h [hunk 7]
```diff
@@ -32,8 +32,15 @@
...
+
 #endif  // ARROW_HAVE_SSE4_2
 
+#ifdef ARROW_HAVE_AVX2
```
Comment (emkornfield) [inline cpp/src/arrow/util/sse_util.h:40]: I think this belongs in the new header you introduced.

**Review by emkornfield — CHANGES_REQUESTED** [review]

File: cpp/src/parquet/encoding.cc [hunk 8]
```diff
@@ -861,8 +861,8 @@ std::shared_ptr<Buffer> ByteStreamSplitEncoder<DType>::FlushValues() {
   uint8_t* output_buffer_raw = output_buffer->mutable_data();
   const size_t num_values = values_.length();
   const uint8_t* raw_values = reinterpret_cast<const uint8_t*>(values_.data());
-#if defined(ARROW_HAVE_SSE4_2)
-  arrow::util::internal::ByteStreamSplitEncodeSSE2<T>(raw_values, num_values,
+#if defined(ARROW_HAVE_SIMD_SPLIT)
```
Comment (emkornfield) [inline cpp/src/parquet/encoding.cc:864]: sorry one more style comment.  We should consolidate this code into byte  byte_stream_split_.h so this code only calls ByteStreamSplitEncode.  Same with decode path below.

File: cpp/src/arrow/util/byte_stream_split_simd_internal.h [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+      const uint8_t byte_in_value = raw_values[i * kNumStreams + j];
+      output_buffer_raw[j * num_values + i] = byte_in_value;
+    }
+  }
+
+  constexpr size_t KNumUnpack = (kNumStreams == 8U) ? 2U : 3U;
+  __m512i final_result[kNumStreams];
+  __m512i unpack[KNumUnpack + 1][kNumStreams];
+  __m512i permutex[kNumStreams];
+  static __m512i permutex_mask;
```
Comment (emkornfield) [inline cpp/src/arrow/util/byte_stream_split_simd_internal.h:471]: does static have an impact here, since this appears to reinitializing each time?  Does the compiler do something reasonable here?

**Review by emkornfield — CHANGES_REQUESTED** [review]
Sorry, some more small code organization comments and one question about the use of static.

I did not take a thorough look at the new SIMD code but the unit tests seam reasonable.  @pitrou or @wesm did you want to take a look at this?

File: cpp/src/parquet/encoding.cc (hunk 8 above)
Comment (jianxind) [reply cpp/src/parquet/encoding.cc]: done

File: cpp/src/arrow/util/byte_stream_split_simd_internal.h (hunk 9 above)
Comment (jianxind) [reply cpp/src/arrow/util/byte_stream_split_simd_internal.h]: Both are optimized to a constant number by the compiler.  _mm512_set_epi32 is a inline helper to pack a _m512i number. Remove the static for a better understanding.

File: cpp/src/arrow/util/byte_stream_split.h [hunk 10]
```diff
@@ -15,10 +15,9 @@
 // specific language governing permissions and limitations
 // under the License.
 
-#ifndef ARROW_UTIL_BYTE_STREAM_SPLIT_H
-#define ARROW_UTIL_BYTE_STREAM_SPLIT_H
+#pragma once
 
-#include "arrow/util/sse_util.h"
+#include "arrow/util/byte_stream_split_simd_internal.h"
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:20]: I don't think there's any reason for a separate header. This one is already specialized enough.

File: cpp/src/arrow/util/byte_stream_split_simd_internal.h [hunk 11]
```diff
@@ -0,0 +1,579 @@
...
+#define ARROW_HAVE_SIMD_SPLIT
+#endif  // ARROW_HAVE_SSE4_2
+
+namespace arrow {
+namespace util {
+namespace internal {
+
+#if defined(ARROW_HAVE_SSE4_2)
+template <typename T>
+void ByteStreamSplitDecodeSse2(const uint8_t* data, int64_t num_values, int64_t stride,
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split_simd_internal.h:36]: Nit: "SSE2"

File: cpp/src/arrow/util/byte_stream_split_simd_internal.h [hunk 12]
```diff
@@ -0,0 +1,579 @@
...
+    for (size_t i = 0; i < kNumStreams; ++i) {
+      _mm_storeu_si128(&output_buffer_streams[i][block_index], final_result[i]);
+    }
+  }
+}
+#endif  // ARROW_HAVE_SSE4_2
+
+#if defined(ARROW_HAVE_AVX2)
+template <typename T>
+void ByteStreamSplitDecodeAvx2(const uint8_t* data, int64_t num_values, int64_t stride,
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split_simd_internal.h:178]: Nit: "AVX2"

File: cpp/src/arrow/util/byte_stream_split_simd_internal.h [hunk 13]
```diff
@@ -0,0 +1,579 @@
...
+    for (size_t i = 0; i < kNumStreams; ++i) {
+      _mm256_storeu_si256(&output_buffer_streams[i][block_index], final_result[i]);
+    }
+  }
+}
+#endif  // ARROW_HAVE_AVX2
+
+#if defined(ARROW_HAVE_AVX512)
+template <typename T>
+void ByteStreamSplitDecodeAvx512(const uint8_t* data, int64_t num_values, int64_t stride,
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split_simd_internal.h:336]: Nit: "AVX512"

File: cpp/src/arrow/util/byte_stream_split_simd_internal.h [hunk 14]
```diff
@@ -0,0 +1,579 @@
...
+  return ByteStreamSplitDecodeAvx2(data, num_values, stride, out);
+#elif defined(ARROW_HAVE_SSE4_2)
+  return ByteStreamSplitDecodeSse2(data, num_values, stride, out);
+#else
+#error "ByteStreamSplitDecodeSimd not implemented"
+#endif
+}
+
+template <typename T>
+void inline ByteStreamSplitEncodeSimd(const uint8_t* raw_values, const size_t num_values,
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split_simd_internal.h:563]: "SIMD"

File: cpp/src/parquet/encoding_benchmark.cc [hunk 15]
```diff
@@ -226,57 +226,30 @@ static void BM_ByteStreamSplitEncode(benchmark::State& state, EncodeFunc&& encod
...
-  BM_ByteStreamSplitEncode<double>(
-      state, arrow::util::internal::ByteStreamSplitEncodeSSE2<double>);
-}
-
-BENCHMARK(BM_ByteStreamSplitDecode_Float_SSE2)->Range(MIN_RANGE, MAX_RANGE);
-BENCHMARK(BM_ByteStreamSplitDecode_Double_SSE2)->Range(MIN_RANGE, MAX_RANGE);
-BENCHMARK(BM_ByteStreamSplitEncode_Float_SSE2)->Range(MIN_RANGE, MAX_RANGE);
-BENCHMARK(BM_ByteStreamSplitEncode_Double_SSE2)->Range(MIN_RANGE, MAX_RANGE);
-#endif
+BENCHMARK(BM_ByteStreamSplitDecode_Float)->Range(MIN_RANGE, MAX_RANGE);
```
Comment (pitrou) [inline cpp/src/parquet/encoding_benchmark.cc:249]: Why did you remove the different implementations? It becomes more difficult to compare them on a given CPU.

File: cpp/src/parquet/encoding_test.cc [hunk 16]
```diff
@@ -1094,16 +1094,26 @@ TYPED_TEST(TestByteStreamSplitEncoding, BasicRoundTrip) {
...
-  // Exercise only the SIMD loop.
-  ASSERT_NO_FATAL_FAILURE(this->Execute(256, 1));
-
-  // Exercise both.
-  ASSERT_NO_FATAL_FAILURE(this->Execute(1337, 1));
-
-  for (int values = 0; values < 32; ++values) {
-    ASSERT_NO_FATAL_FAILURE(this->Execute(values, 1));
-  }
+  constexpr size_t kSuffixSize = 7;
```
Comment (pitrou) [inline cpp/src/parquet/encoding_test.cc:1097]: I'm not convinced all this is better than a loop over a range of possible sizes as above.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you. The speed improvements are nice, though this is not a very frequent code path.

**Review by emkornfield — COMMENTED** [review]
@pitrou I think I recommended the separate header for Sunday implementation s to make it easier to track but I'm ok combining.  

For naming the [style guide]( https://google.github.io/styleguide/cppguide.html#General_Naming_Rules) dictates acronyms should be camel cased.

**Comment by pitrou** [comment]
> For naming the style guide dictates acronyms should be camel cased.

Ah, fair enough.

**Review by pitrou — APPROVED** [review]
Looks fine to me now, thank you.
@emkornfield do you want to do another pass?

File: cpp/src/arrow/util/byte_stream_split.h [hunk 17]
```diff
@@ -163,21 +171,421 @@ void ByteStreamSplitEncodeSSE2(const uint8_t* raw_values, const size_t num_value
...
+    for (size_t i = 0; i < kNumStreams; ++i) {
       _mm_storeu_si128(&output_buffer_streams[i][block_index], final_result[i]);
     }
   }
 }
+#endif  // ARROW_HAVE_SSE4_2
+
+#if defined(ARROW_HAVE_AVX2)
+template <typename T>
+void ByteStreamSplitDecodeAVX2(const uint8_t* data, int64_t num_values, int64_t stride,
```
Comment (emkornfield) [inline cpp/src/arrow/util/byte_stream_split.h:183]: sorry for the churn, I think lets change these back to Avx/Simd, etc (see my reply to @pitrou about this being more consistent with the style guide).

**Review by emkornfield — CHANGES_REQUESTED** [review]
Other then the capitalization issue LGTM.

File: cpp/src/arrow/util/byte_stream_split.h (hunk 17 above)
Comment (jianxind) [reply cpp/src/arrow/util/byte_stream_split.h]: Thanks. Changed to camel cased per the coding guide.

**Comment by pitrou** [comment]
Thank you for being persistent @jianxind :-)

**Comment by pitrou** [comment]
For the record, here are the results on an AMD Ryzen with clang 9.0:
```
-------------------------------------------------------------------------------------------------------
Benchmark                                             Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------------------------
BM_ByteStreamSplitDecode_Float_Scalar/1024         5602 ns         5601 ns       120412 bytes_per_second=697.445M/s
BM_ByteStreamSplitDecode_Float_Scalar/4096        22393 ns        22391 ns        31258 bytes_per_second=697.835M/s
BM_ByteStreamSplitDecode_Float_Scalar/32768      179862 ns       179838 ns         3892 bytes_per_second=695.071M/s
BM_ByteStreamSplitDecode_Float_Scalar/65536      361241 ns       361193 ns         1936 bytes_per_second=692.152M/s
BM_ByteStreamSplitDecode_Double_Scalar/1024        5641 ns         5640 ns       123423 bytes_per_second=1.35274G/s
BM_ByteStreamSplitDecode_Double_Scalar/4096       22747 ns        22744 ns        30821 bytes_per_second=1.3418G/s
BM_ByteStreamSplitDecode_Double_Scalar/32768     182204 ns       182179 ns         3843 bytes_per_second=1.34012G/s
BM_ByteStreamSplitDecode_Double_Scalar/65536     364179 ns       364133 ns         1922 bytes_per_second=1.34094G/s
BM_ByteStreamSplitEncode_Float_Scalar/1024         1190 ns         1190 ns       588888 bytes_per_second=3.20508G/s
BM_ByteStreamSplitEncode_Float_Scalar/4096         4753 ns         4753 ns       147221 bytes_per_second=3.21053G/s
BM_ByteStreamSplitEncode_Float_Scalar/32768       38483 ns        38479 ns        18179 bytes_per_second=3.17243G/s
BM_ByteStreamSplitEncode_Float_Scalar/65536       77039 ns        77028 ns         9081 bytes_per_second=3.16952G/s
BM_ByteStreamSplitEncode_Double_Scalar/1024        2362 ns         2362 ns       293478 bytes_per_second=3.23025G/s
BM_ByteStreamSplitEncode_Double_Scalar/4096       10302 ns        10301 ns        67743 bytes_per_second=2.9626G/s
BM_ByteStreamSplitEncode_Double_Scalar/32768      79843 ns        79833 ns         8757 bytes_per_second=3.05816G/s
BM_ByteStreamSplitEncode_Double_Scalar/65536     159168 ns       159147 ns         4386 bytes_per_second=3.06812G/s
BM_ByteStreamSplitDecode_Float_Sse2/1024            428 ns          428 ns      1633944 bytes_per_second=8.91636G/s
BM_ByteStreamSplitDecode_Float_Sse2/4096           1581 ns         1581 ns       444223 bytes_per_second=9.65414G/s
BM_ByteStreamSplitDecode_Float_Sse2/32768         12606 ns        12604 ns        55404 bytes_per_second=9.68487G/s
BM_ByteStreamSplitDecode_Float_Sse2/65536         25412 ns        25408 ns        27542 bytes_per_second=9.60867G/s
BM_ByteStreamSplitDecode_Double_Sse2/1024           749 ns          749 ns       946357 bytes_per_second=10.1817G/s
BM_ByteStreamSplitDecode_Double_Sse2/4096          2886 ns         2885 ns       244727 bytes_per_second=10.5778G/s
BM_ByteStreamSplitDecode_Double_Sse2/32768        23120 ns        23117 ns        30184 bytes_per_second=10.561G/s
BM_ByteStreamSplitDecode_Double_Sse2/65536        46524 ns        46517 ns        15028 bytes_per_second=10.4968G/s
BM_ByteStreamSplitEncode_Float_Sse2/1024            226 ns          226 ns      3092645 bytes_per_second=16.8935G/s
BM_ByteStreamSplitEncode_Float_Sse2/4096            905 ns          905 ns       774657 bytes_per_second=16.8537G/s
BM_ByteStreamSplitEncode_Float_Sse2/32768          7253 ns         7252 ns        95973 bytes_per_second=16.8327G/s
BM_ByteStreamSplitEncode_Float_Sse2/65536         14367 ns        14365 ns        48506 bytes_per_second=16.995G/s
BM_ByteStreamSplitEncode_Double_Sse2/1024           632 ns          632 ns      1106593 bytes_per_second=12.0722G/s
BM_ByteStreamSplitEncode_Double_Sse2/4096          2641 ns         2641 ns       262638 bytes_per_second=11.5554G/s
BM_ByteStreamSplitEncode_Double_Sse2/32768        21205 ns        21201 ns        32961 bytes_per_second=11.5153G/s
BM_ByteStreamSplitEncode_Double_Sse2/65536        43233 ns        43227 ns        16301 bytes_per_second=11.2958G/s
BM_ByteStreamSplitDecode_Float_Avx2/1024            102 ns          102 ns      6852810 bytes_per_second=37.4232G/s
BM_ByteStreamSplitDecode_Float_Avx2/4096            367 ns          367 ns      1908784 bytes_per_second=41.5411G/s
BM_ByteStreamSplitDecode_Float_Avx2/32768          2927 ns         2927 ns       239104 bytes_per_second=41.7092G/s
BM_ByteStreamSplitDecode_Float_Avx2/65536          6093 ns         6092 ns       114715 bytes_per_second=40.0731G/s
BM_ByteStreamSplitDecode_Double_Avx2/1024           196 ns          196 ns      3495638 bytes_per_second=38.9612G/s
BM_ByteStreamSplitDecode_Double_Avx2/4096           935 ns          935 ns       739466 bytes_per_second=32.6352G/s
BM_ByteStreamSplitDecode_Double_Avx2/32768         7656 ns         7655 ns        90728 bytes_per_second=31.8937G/s
BM_ByteStreamSplitDecode_Double_Avx2/65536        16258 ns        16255 ns        43471 bytes_per_second=30.0381G/s
BM_ByteStreamSplitEncode_Float_Avx2/1024            153 ns          153 ns      4562139 bytes_per_second=24.9177G/s
BM_ByteStreamSplitEncode_Float_Avx2/4096            611 ns          611 ns      1137500 bytes_per_second=24.9915G/s
BM_ByteStreamSplitEncode_Float_Avx2/32768          4884 ns         4884 ns       143125 bytes_per_second=24.9958G/s
BM_ByteStreamSplitEncode_Float_Avx2/65536          9827 ns         9826 ns        70977 bytes_per_second=24.8474G/s
BM_ByteStreamSplitEncode_Double_Avx2/1024           626 ns          625 ns      1114075 bytes_per_second=12.1979G/s
BM_ByteStreamSplitEncode_Double_Avx2/4096          2638 ns         2638 ns       263590 bytes_per_second=11.5693G/s
BM_ByteStreamSplitEncode_Double_Avx2/32768        21236 ns        21233 ns        32894 bytes_per_second=11.4982G/s
BM_ByteStreamSplitEncode_Double_Avx2/65536        42853 ns        42845 ns        16381 bytes_per_second=11.3964G/s
```

**Comment by pitrou** [comment]
@jianxind Feel free to ask if you want to look for other performance work.

**Comment by cyb70289** [comment]
Post an interesting finding: gcc is able to auto vectorize the naive scalar implementation and generate much faster code than clang, approaching handwritten sse4.2 performance. With gcc, the benchmark result is kind of weird, sse4 is faster than avx2, maybe longer test data will fix the gap.
See https://godbolt.org/z/joQ4iy.

**Comment by jianxind** [comment]
> Post an interesting finding: gcc is able to auto vectorize the naive scalar implementation and generate much faster code than clang, approaching handwritten sse4.2 performance(Encode_Float test). With gcc, the benchmark result is kind of weird, sse4 is faster than avx2, maybe longer test data will fix the gap.
> See https://godbolt.org/z/joQ4iy.

Ahh, which path for bad AVX2 result, decode or encode, float or double, and size? The data I post was measured on a server which support AVX512, and I also has a i9 client which capable of AVX2, all result shows AVX2 better than SSE. Maybe the behavior is different per HW, but generally say we can achieve more benefit from the advanced  SIMD instructions.

**Comment by cyb70289** [comment]
> Ahh, which path for bad AVX2 result, decode or encode, float or double, and size? The data I post was measured on a server which support AVX512, and I also has a i9 client which capable of AVX2, all result shows AVX2 better than SSE. Maybe the behavior is different per HW, but generally say we can achieve more benefit from the advanced SIMD instructions.

Sorry I forgot to say my test env, which is essential to make sense benchmark results.
I tested on AMD EYPC 7251 (znver1), build with gcc-7.4 and clang-9.
Benchmark result is in this link: https://pastebin.com/g1b9C6Tp
With gcc, sse4 code is faster than avx2 on that platform.
I tried some Intel platforms(i7, e5) just now, they work as expected, avx2 > sse4 >> scalar.

This is just some random findings, doesn't mean anything concrete. For performance critical code, I think hand written optimization code has much more consistent and predictable behaviour than compiler generated ones. Some projects (e.g, intel [isa-l](https://github.com/intel/isa-l)) go extreme to write everything with assembly to bypass compiler optimizer completely.

**Comment by pitrou** [comment]
AMD "Zen" CPUs apparently [have 128-bit SIMD units](https://www.anandtech.com/show/11170/the-amd-zen-and-ryzen-7-review-a-deep-dive-on-1800x-1700x-and-1700/8): 256-bit AVX instructions are split in two instructions for execution.

AMD "Zen 2" CPUs, however, [have 256-bit SIMD units](https://www.anandtech.com/show/14525/amd-zen-2-microarchitecture-analysis-ryzen-3000-and-epyc-rome/9), and they benefit from AVX code: see the numbers I posted above (I have a AMD Ryzen 9 3900X).

**Comment by cyb70289** [comment]
Make sense. Thank you.
